ext/gctrack/gctrack.c

```diff
@@ -67,14 +89,28 @@ gctracker_hook(VALUE tpval, void *data)
       last_enter = 0;
     }
       break;
+    case RUBY_INTERNAL_EVENT_NEWOBJ: {
+      if (gctracker_enabled() && last_record) {
+
+        add_allocations();
+      }
+    }
+      break;
+    case RUBY_INTERNAL_EVENT_SWITCH: {
```

Comment:
While I get the metric, and it probably does make sense, wouldn't we really care about `ru_(n)vcsw` data from `getrusage(2)` ?

Reply:
Ah, yeah, interesting point. I suppose that Ruby threads are always real threads these days. I wonder whether they'd ever be different? I guess if something else works on the CPU it'd get incremented as well, whereas this tracepoint wouldn't?

The reason I wanted this was to get a feel for which requests do context switches in Puma.

Reply:
Well, `RUBY_INTERNAL_EVENT_SWITCH` can't possibly track non-voluntary ones. For voluntary ones, as long as the thread switched in is a ruby thread, I _should_ be reflected in `RUBY_INTERNAL_EVENT_SWITCH`, I think so too (modulo missing code paths). Now `RUBY_INTERNAL_EVENT_SWITCH` is a VM global event, so you'd need to do the math yourself. You should be able to get access to all the `rb_thread_*` api for when event is fired. `getrusage(2)` has support for `RUSAGE_THREAD`, i.e. stats about the calling thread,  which would help answer the question "which requests do context switches" if indeed all threads in Puma match to native ones. 
Hope this clarifies my thinking somewhat, but there is lot to unpack here I fear, sorry!
I reached out to @felix-d and we can then sync IRL - which would make it a lot easier.

Reply:
> The reason I wanted this was to get a feel for which requests do context switches in Puma.

I think that this will report what Simon is looking for. While Ruby threads _are_ real pthreads (on Linux), the GVL will limit Ruby code to running on a single thread at a time. Looking at the code, it appears that `RUBY_INTERNAL_EVENT_SWITCH` is indirectly called by the [timer thread](https://github.com/ruby/ruby/blob/v2_5_3/thread_pthread.c#L1552-L1590). As far as I know, the Ruby timer thread is what's used to ensure fair scheduling among Ruby threads, and ensure that no single Ruby thread can monopolize the GVL. I'm assuming that Puma threads are implemented as Ruby threads, and not using raw pthreads through something like a C extension.

Reply:
Reading this code a little further, it seems that's only timer triggered indeed. So that's good as, if I remember this right, other blocking calls, as IO, wouldn't trigger an increment. Which doesn't mean you have perfect scheduling neither, but possibly helps identifying CPU intensive code blocks. Still, I find the semantic of the value hard to reason about. Especially as that number would still be for the whole VM, not a per thread basis, i.e. for any value different than 2, you don't know that happened, right? 
i.e. you get scheduled out, and another thread gets to make progress, than that's being scheduled out for you to come back in; so that's 2. But if you got scheduled in and out a couple of times, or whether 4 other threads got scheduled in until you got CPU time again, you'd need to correlate the data point (over that period) for all threads/requests. Not that is necessarily an issue, just pointing out that it isn't equivalent to the `RUSAGE_THREAD` flag you could use with `getrusage(2)`. But I can see this isn't what's desired here too. 
In either case, we'd need to make this thread aware first and then (possibly) count per thread and/or apply data to relevant thread-local counters. Finally, possibly having a global value for the process for involuntary ones, could be useful too, to help identify over-commissioned cases, but that's probably out of scope.

wrt to the threading model, I implied model 2 from [thread.c](https://github.com/ruby/ruby/blob/v2_5_3/thread.c#L31-L40), but … probably not generically safe indeed.